### ISSUE TITLE
get_mesh_laplacian: summation row-wise is non-zero

### ISSUE THREAD

**Issue by xieweiyi** [pyg-team/pytorch_geometric#6037]
#4187  

row-wise summation of the laplacian matrix computed by get_mesh_laplacian is non-zero.

```
from torch_geometric.data import Data
from torch_geometric.utils import get_mesh_laplacian
face = torch.tensor([[0, 2, 3, 0], [1, 1, 2, 2], [2, 4, 4, 3]], dtype=torch.long)
pos = torch.tensor([[0, 1, 0], [1, 0, 0], [1, 1, 0], [1, 2, 0], [2, 1, 0]], dtype=torch.float)

data = Data(pos=pos, face=face)
edge_index, edge_weights = get_mesh_laplacian(data.pos, data.face)
row, col = edge_index
row_wise_sum = scatter_add(edge_weights, row, dim=0, dim_size=5)
assert torch.all(row_wise_sum == 0), "row_wise summation is non_zero"
```

### Environment

* PyG version: 2.1.0.post1
* PyTorch version: 1.12.1
* OS: windows 11
* Python version: 3.9
* CUDA/cuDNN version: cpu
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* torch-scatter: 2.0.9
* torch-sparse: 0.6.15

**Comment by rusty1s** [comment]
Sorry for late reply. Do you have some pointers on the bug here?

**Comment by xieweiyi** [comment]
I think it is sometimes preferred to have the Laplacian normalized such that the sum of each row/col is 1.0.

**Comment by rusty1s** [comment]
We can add in a normalized option. Would that resolve your concerns?

**Comment by xieweiyi** [comment]
Yes, thanks, that would be great.

**Comment by daniel-unyi-42** [comment]
Thanks for pointing this out, and apologies for the incorrect implementation!

The definition of the cotangent Laplacian is $L_{ij} = \frac{\cot \angle_{ikj} + \cot \angle_{ilj}}{2}$ if there's an edge between $i$ and $j$, $w_{ij}=0$ if there's no edge between $i$ and $j$, and $w_{ij}=\sum_{j \in N(i)}{L_{ij}}$ in the diagonal. The cotangent Laplacian has many desirable properties, e.g. it's proven to converge to the continuous Laplace-Beltrami operator. However, the cotangent weights depend on mesh sampling, so people sometimes modify it using the diagonal mass matrix $M$. A good summary can be found in [Discrete Laplace–Beltrami operators for shape analysis and segmentation by Reuter et al.](https://www.sciencedirect.com/science/article/pii/S0097849309000272):
- $M_{ii}=1$, this is the cotangent Laplacian
- $M_{ii}=a(i)/3$, where $a(i)$ is the area obtained by joining the centroids of the triangles around vertex $i$
- $M_{ii}=a_V(i)$, where $a_V(i)$ is the area obtained by joining the circumcenters of the triangles around vertex $i$

Using the mass matrix, we can normalize the cotangent Laplacian row-wise $M^{-1}L$ or symmetrically $M^{-1/2}LM^{-1/2}$.
In each case, the row-wise summation of the mesh Laplacian should be zero.

I'm going to correct the implementation, and also extend it, such that the user can choose between no normalization, row-wise normalization and symmetric normalization (using the centroid area as mass).
